### ISSUE TITLE
Using name mangling on a method parameter suddenly makes a protocol work

### ISSUE THREAD

**Issue by samvv** [microsoft/pyright#4969]
**Describe the bug**

_Not sure this is a bug or something I get wrong, but I thought I'd let it know nonetheless. This issue might need to be moved over to the MyPy repository but I thought I'd start here._

pyright only considers a specific signature to be valid when prefixing one of the parameters with `__`, which seems weird and unexpected. Name mangling does not magically hide a parameter, so I'm not sure what's going on here.

**To Reproduce**

 - Install pyright 1.1.303
 - Write the snippet at the end of this description to a file.
 - Run pyright on the source file and watch as it reports an error about assiging `int` with the `__lt__` member.
 - Now replace parameter name `value` with `__value` inside the `__lt__`-method.
 - Watch it run successfully, as should be the case with the first snippet too.

**Expected behavior**

As far as my understanding of Python goes, both snippets should be equal and produce no typing error.

**Screenshots or Code**

This doesn't work on my machine:
```python
from typing import Protocol, Any

class Comparable(Protocol):
  def __lt__(self, value: Any) -> bool: ...

foo: Comparable = int(1)
```

```
sorting.py
  sorting.py:10:19 - error: Expression of type "int" cannot be assigned to declared type "Comparable"
    "int" is incompatible with protocol "Comparable"
      "__lt__" is an incompatible type
        Type "(__value: int, /) -> bool" cannot be assigned to type "(value: int) -> bool"
          Position-only parameter mismatch; expected 1 but received 0 (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.759sec
```

**VS Code extension or command-line**

Using Neovim with LSP configured to pyright version 1.1.303.
Also tried the command-line directly with as only argument the source file to be checked.

**Comment by samvv** [comment]
Relevant type definition of the `int` comparator method in typeshed: https://github.com/python/mypy/blob/0799a8ab0dc8deed8d2e0ec34b1aab2fe39ebd96/mypy/typeshed/stdlib/builtins.pyi#L295

**Comment by erictraut** [comment]
Pyright is working as expected here. The protocol indicates that the `__lt__` method must accept a parameter named `value` and that this parameter can be specified by keyword, but the `int` class doesn't provide a compatible method `__lt__`. The `int.__lt__` method doesn't accept any keyword parameters; all of its parameters are positional-only. 

[PEP 570](https://peps.python.org/pep-0570) introduced a proper way to specify positional-only parameters in a signature using the `/` marker. You can place this marker at the end of a parameter list to indicate that all parameters are positional-only.

```python
class Comparable(Protocol):
    def __lt__(self, value: Any, /) -> bool:
        ...
```

Prior to Python 3.8, the `/` marker was not allowed. For older versions, a naming convention was used instead. Parameters that are named starting with a `__` are considered positional-only. Unlike class attributes, these parameter names are not mangled. It's simply a convention that tells static type checkers that it should treat the parameter as positional-only. 

If you're not using Python 3.7 or older, I recommend using the PEP 570 notation instead.

**Comment by samvv** [comment]
Thank you very much for explaining that to me @erictraut ! Now it makes sense.

Good luck with the project.
